vingo/src/routes/scans.rs

```diff
@@ -102,3 +106,40 @@ pub async fn add(state: State<AppState>, body: String) -> ResponseResult<String>
 
     Ok(user.name)
 }
+
+#[derive(Debug, FromQueryResult, Serialize)]
+pub struct RecentScanItem {
+    id: i32,
+    scan_time: DateTimeWithTimeZone,
+}
+
+pub async fn recent(state: State<AppState>) -> ResponseResult<Json<Vec<RecentScanItem>>> {
+    let fourteen_days_ago = (Local::now() - Duration::days(14)).naive_local();
+    let scans = Scan::find()
+        .select_only()
+        .expr(Expr::cust(format!(
+            "DISTINCT ON ({}, DATE({})) scan.{}, scan.{}", //NOTE: this is a pain
+            card::Column::UserId.as_str(),
+            scan::Column::ScanTime.as_str(),
+            scan::Column::Id.as_str(),
+            scan::Column::ScanTime.as_str(),
+        )))
+        .join(JoinType::InnerJoin, scan::Relation::Card.def())
+        .join(JoinType::InnerJoin, card::Relation::User.def())
+        .filter(scan::Column::ScanTime.gt(fourteen_days_ago))
+        .order_by_asc(card::Column::UserId)
```

Comment:
Door te sorteren op (userid, scantime, scan id) zorg je ervoor dat je vrij goed in de Zeus-publieke output de scans van een bepaalde persoon aan elkaar gaat kunnen linken (die gaan namelijk sequentieel in de output staan). Als je dan bijvoorbeeld iemand ziet scannen op cammie en dat linkt aan de tijd, weet je ook wanneer die persoon de afgelopen 14 dagen gescanned heeft. Als die persoon vaker dan 1 keer per 14 dagen scant, dan kan je dat (mits periodiek scrapen) zelfs tot ver terug in de tijd linken.

Het lijkt me beter om de output te sorteren op enkel scan datetime (want die zit sowieso in de output, en zo weet je zeker dat je geen impliciete sortering van je database exposet).

Reply:
goed punt